### ISSUE TITLE
batch size confusion

### ISSUE THREAD

**Issue by Moldoteck** [microsoft/DeepSpeed#1082]
Hi. I am a bit confused: What happen when I have an torch.utils.Dataloader with batch size = 7 BUT in the deepspeed config i have set>
```
"gradient_accumulation_steps": 1,
"train_micro_batch_size_per_gpu": 20,
```

P.S. I am performing training on single GPU P100

**Comment by tjruwase** [comment]
In this case, DeepSpeed will report a larger samples per second statistics (assuming micro batch size of 20 instead of 7). But it should not have any impact on training correctness.

What are you observing?

**Comment by Moldoteck** [comment]
> 
> 
> In this case, DeepSpeed will report a larger samples per second statistics (assuming micro batch size of 20 instead of 7). But it should not have any impact on training correctness.
> 
> What are you observing?

Just asked out of curiosity. Thank you for your response and by the way, your framework is very cool!

Also, if you have some time to explain:
I have set up parameter offloading and optimizer offloading for stage 3 in json for my model _(6M params), single gpu P100_, but it doesn't help increasing batch size: my original max batch size is 7 and even with offloading i get OOM for batch size bigger than 7. Could this be because of some specific architecture/dataset properties or parameter+optimizer offloading alone do not give big memory savings? P.S. here is the config i use:
```
{
    "gradient_accumulation_steps": 1,
    "train_micro_batch_size_per_gpu": 10,
    "gradient_clipping":1.0,
    "zero_optimization": {
        "stage": 3,
        "reduce_bucket_size": 2e6,
        "allgather_bucket_size": 2e6,
        "stage3_max_live_parameters": 1e7,
        "stage3_max_reuse_distance": 1e7,
        "stage3_prefetch_bucket_size": 5e6,
        "stage3_param_persistence_threshold": 1e5,
        "contiguous_gradients": true,
        "overlap_comm": false,
        "offload_param": {
            "device": "cpu",
            "pin_memory": true
        },
        "offload_optimizer": {
            "device": "cpu",
            "pin_memory": true
        }
    }
}
```
And this is the initialization part:
```
...
with torch.no_grad():
    self.model = RandomModel()

if torch.cuda.is_available() and torch.cuda.device_count() > 0:
    self.model.cuda()

self.optimizer = deepspeed.ops.adam.DeepSpeedCPUAdam(self.model.parameters(), lr=0.01, bias_correction=True, betas=(0.9, 0.999), eps=1e-08, weight_decay=0.0001, amsgrad=False, adamw_mode=True)
self.scheduler = deepspeed.runtime.lr_schedules.WarmupDecayLR(optimizer= self.optimizer, total_num_steps = 300, warmup_min_lr = 0.0, warmup_max_lr = 0.01, warmup_num_steps = 2, last_batch_iteration = -1)
self.model, _, _, self.scheduler = deepspeed.initialize(model=self.model, optimizer=self.optimizer, lr_scheduler=self.scheduler, model_parameters=self.model.parameters(), config='./modules/ds_config.json')
...
```
And this is training part
```
...
output = model(in_vol)
loss_m = someloss(output)
model.backward(loss_m)
model.step()
...
```

**Comment by tjruwase** [comment]
A 6M model should easily fit into 12GB of a single P100. This suggests OOM is due to something other than model size. ZeRO is most useful when model size is the bottleneck.  A few things to try?

1. Share a stack trace of the OOM
2. Use fp16 mixed-precision training
3. Take a look at #467

**Comment by tjruwase** [comment]
@Moldoteck, is this still an issue?

**Comment by Moldoteck** [comment]
> 
> 
> @Moldoteck, is this still an issue?

Thanks for asking. Main improvement where from model checkpoint and CPU activations offload. Other things didn't help too much. The fp16 training produced exploding gradients. Anyway, that's not part of the ticket, so it can be closed

**Comment by tjruwase** [comment]
Thanks for the confirmation.
